### PR TITLE
Add WithStatus option to set Span status at the same time (#1677)

### PR DESCRIPTION
**Pull request by lastchiliarch**
Add WithStatus event option to set Span status at the same time, see #1677.
Signed-off-by: lastchiliarch <lastchiliarch@163.com>

### CODE REVIEW

File: sdk/trace/trace_test.go [hunk 1]
```diff
@@ -1164,6 +1164,59 @@ func TestRecordErrorNil(t *testing.T) {
...
+func TestRecordErrorWithStatus(t *testing.T) {
+	scenarios := []struct {
+		err error
+		typ string
+		msg string
+	}{
+		{
+			err: ottest.NewTestError("test error"),
+			typ: "go.opentelemetry.io/otel/internal/internaltest.TestError",
+			msg: "test error",
```
Comment (XSAM) [inline sdk/trace/trace_test.go:1176]: Should we add a test case that the error is nil?

File: sdk/trace/span.go [hunk 2]
```diff
@@ -244,14 +244,18 @@ func (s *span) End(options ...trace.SpanOption) {
...
 }
 
-// RecordError will record err as a span event for this span. An additional call to
-// SetStatus is required if the Status of the Span should be set to Error, this method
-// does not change the Span status. If this span is not being recorded or err is nil
-// than this method does nothing.
+// RecordError will record err as a span event for this span.
+// this method does not change the Span status in default.
+// If you want to change Span status, pass WithStatus as opts.
+// this metod does nothing If this span is not being recorded or err is nil.
```
Comment (XSAM) [inline sdk/trace/span.go:250]:
```suggestion
// This method does not change the Span status in default.
// To change Span status, pass `WithStatus` as options.
// This method does nothing If this span is not being recorded or err is nil.
```

File: trace/config_test.go [hunk 3]
```diff
@@ -40,6 +40,10 @@ func TestNewSpanConfig(t *testing.T) {
 		Attributes:  []attribute.KeyValue{k1v2, k2v2},
 	}
 
+	withStatusOpt := WithStatus(true)
+	//just for coverage
```
Comment (XSAM) [inline trace/config_test.go:44]:
```suggestion
	// For coverage
```

**Review by XSAM — APPROVED** [review]
Looks good overall.

File: sdk/trace/span.go (hunk 2 above)
Comment (lastchiliarch) [reply sdk/trace/span.go]: Thanks a lot.

File: sdk/trace/trace_test.go (hunk 1 above)
Comment (lastchiliarch) [reply sdk/trace/trace_test.go]: Done, thank you very much.

File: sdk/trace/span.go [hunk 4]
```diff
@@ -244,14 +244,18 @@ func (s *span) End(options ...trace.SpanOption) {
...
 }
 
-// RecordError will record err as a span event for this span. An additional call to
-// SetStatus is required if the Status of the Span should be set to Error, this method
-// does not change the Span status. If this span is not being recorded or err is nil
-// than this method does nothing.
+// RecordError will record err as a span event for this span.
+// This method does not change the Span status in default
+// To change Span status, pass `WithStatus` as options.
+// This method does nothing If this span is not being recorded or err is nil.
```
Comment (MrAlias) [inline sdk/trace/span.go:250]:
```suggestion
// This method does not change the Span status by default.
// To set the Span status, pass `WithStatus` as an option.
// This method does nothing if this span is not being recorded or err is nil.
```

**Review by MrAlias — CHANGES_REQUESTED** [review]

File: trace/config.go [hunk 5]
```diff
@@ -203,3 +205,15 @@ func (i instrumentationVersionOption) ApplyTracer(config *TracerConfig) {
...
+
+type withStatusSpanOption bool
+
+func (o withStatusSpanOption) ApplySpan(c *SpanConfig)  { o.apply(c) }
+func (o withStatusSpanOption) ApplyEvent(c *SpanConfig) { o.apply(c) }
+func (withStatusSpanOption) private()                   {}
+func (o withStatusSpanOption) apply(c *SpanConfig)      { c.WithStatus = bool(o) }
+
+// WithStatus set the status at the same time when RecordError
+func WithStatus(o bool) LifeCycleOption {
```
[the remark this answers is not in the document]
Comment (Aneurysm9) [reply trace/config.go:217]: > Additionally, having this accept a boolean as an argument doesn't seem that useful. Why not accept a code and message for the status to set?

I would second that, this at the minimum should accept a code and message as the span status concept has room to grow beyond the current tri-state values.

[the remark this answers is not in the document]
Comment (paivagustavo) [reply trace/config.go:217]: I'm not sure about this either. `WithStatus`doesn't seems to be an `EventOption`, it does not change or affect an `Event`. It is only applicable to the `RecordError` method.

For example, using `span.AddEvent("some event", trace.WithStatus())` is syntactically correct but does nothing. 

I would prefer if we create another method to record an error with status or create an `ErrorOption` that wraps an `EventOptions`, for example:
``` go
span.RecordErrorStatus(err, code, "description", EventOptions..)

```

or

``` go
span.RecordError(err, trace.WithErrorStatus(code, "description"), trace.WithEventOpts(EventOptions..))
```

[the remark this answers is not in the document]
Comment (MrAlias) [reply trace/config.go:217]: I like the creation of the `ErrorOption` approach. I could see there being many more things we want to pass here (we talked about maybe an option that would signal the stacktrace needed to be recorded).

[the remark this answers is not in the document]
Comment (Aneurysm9) [reply trace/config.go:217]: [What is dead](https://github.com/open-telemetry/opentelemetry-go/commit/ec300b28ada49f907c9e65b8140573fa441eba69#diff-42cef9f943e3e301453d0258f416787fdbeaf98a3d6b26bfefb237730d8f25acL255)
[May never die](https://github.com/open-telemetry/opentelemetry-go/commit/ca3f74d9769c4e32310ddb6538b77c4c8bafe7e0#diff-e7647a765bb5a08a1704b2d21a10868009b614a20c9b577fcab32959452f4b89R79-R84)

[the remark this answers is not in the document]
Comment (lastchiliarch) [reply trace/config.go:217]: ErrorOption sounds good. I will use  ErrorOption to replace WithStatusOption.
What can I do with current pr, close it  and open a new one or just work on this?

File: trace/config.go [hunk 6]
```diff
@@ -203,3 +205,56 @@ func (i instrumentationVersionOption) ApplyTracer(config *TracerConfig) {
...
+	// violate compatibility.
+	private()
+}
+
+// NewErrorConfig applies all the options to a returned ErrorConfig.
+func NewErrorConfig(options ...ErrorOption) *ErrorConfig {
+	c := new(ErrorConfig)
+	for _, option := range options {
+		option.ApplyError(c)
+	}
```
Comment (Aneurysm9) [inline trace/config.go:232]: I'd like to avoid the need to use `WithEventOpts()` when calling `RecordError()`.  My first thought was that maybe something like this would work: 

 ```go
	for _, option := range options {
		if evopt, ok := option.(EventOption); ok {
			c.EventOpts = append(c.EventOpts, option)
		} else {
			option.ApplyError(c)
		}
	}
```

But I don't think it does since the method signature now requires `ErrorOption`s to be passed instead of `EventOption`s.  So, my next thought is could the relevant `EventOption` implementations gain an `ErrorOption` implementation that looks something like this:

```go
func (o someOptionType) ApplyError(c *ErrorConfig) {
	c.EventOpts = append(c.EventOpts, o)
}
```

File: trace/config.go [hunk 7]
```diff
@@ -203,3 +205,56 @@ func (i instrumentationVersionOption) ApplyTracer(config *TracerConfig) {
...
+	return statusErrorOption{code, message}
+}
+
+type eventOptsErrorOption []EventOption
+
+func (o eventOptsErrorOption) ApplyError(c *ErrorConfig) { c.EventOpts = o }
+func (eventOptsErrorOption) private()                    {}
+
+// WithEventOpts set event options, will be passed to addEvent
+func WithEventOpts(option ...EventOption) ErrorOption {
```
Comment (Aneurysm9) [inline trace/config.go:258]:
```suggestion
func WithEventOptions(option ...EventOption) ErrorOption {
```

If we can't get rid of this I'd like it to not use an abbreviated name.

File: trace/trace.go [hunk 8]
```diff
@@ -515,11 +515,11 @@ type Span interface {
...
 
-	// RecordError will record err as an exception span event for this span. An
-	// additional call toSetStatus is required if the Status of the Span should
-	// be set to Error, this method does not change the Span status. If this
-	// span is not being recorded or err is nil than this method does nothing.
-	RecordError(err error, options ...EventOption)
+	// RecordError will record err as an exception span event for this span.
+	// This method does not change the Span status by default.
+	// To set the Span status, pass `WithErrorStatus` as an option.
+	// This method does nothing if this span is not being recorded or err is nil.
```
Comment (Aneurysm9) [inline trace/trace.go:521]:
```suggestion
	// This method does not change the Span status by default
	// To set the Span status pass a `WithErrorStatus(status)` option.
	// This method does nothing if this span is not being recorded or if err is nil.
```

**Review by Aneurysm9 — CHANGES_REQUESTED** [review]
I'd really like to avoid the extra `WithEventOptions()` wrapper.

File: trace/config.go (hunk 6 above)
Comment (lastchiliarch) [reply trace/config.go:232]: I don't like the wrap either,  I had tried this  way to  make it work before the wrap.

But it can't work, beacuse the compile complains about the type mismatch.
Error message:
     Cannot use 'trace.WithTimestamp(errTime)' (type LifeCycleOption) as the type ErrorOption Type does not implement 
   'ErrorOption' as some methods are missing: ApplyError(config *ErrorConfig)

Code looks like this:

```
config:
 type timestampSpanOption time.Time

func (o timestampSpanOption) ApplySpan(c *SpanConfig)  { o.apply(c) }
func (o timestampSpanOption) ApplyEvent(c *SpanConfig) { o.apply(c) }
func (o timestampSpanOption) ApplyError(c *ErrorConfig) {c.Timestamp = time.Time(o)}
func (timestampSpanOption) private()                   {}
func (o timestampSpanOption) apply(c *SpanConfig)      { c.Timestamp = time.Time(o) }



client code:
span.RecordError(s.err, trace.WithTimestamp(errTime))


RecordError:

func (s *span) RecordError(err error, options ...trace.ErrorOption) {
	if s == nil || err == nil || !s.IsRecording() {
		return
	}
	c := trace.NewErrorConfig(options...)
	if c.Code != codes.Unset {
		s.SetStatus(c.Code, c.Message)
	}

	var eventOpts []trace.EventOption
	for _, option := range options {
		if option, ok := option.(trace.EventOption); ok {
			eventOpts = append(eventOpts, option)
		}
	}
	eventOpts = append(eventOpts, trace.WithAttributes(
		semconv.ExceptionTypeKey.String(typeStr(err)),
		semconv.ExceptionMessageKey.String(err.Error()),
	))
	s.addEvent(semconv.ExceptionEventName, eventOpts...)
}

```



I don't think it's a good idea to make LifeCycleOption implements ErrorOption.

If you know any idea, that will be a great help, I'd like to remove the wrap.

File: trace/config.go (hunk 7 above)
Comment (lastchiliarch) [reply trace/config.go:258]: Thanks very much.

File: trace/config.go [hunk 9]
```diff
@@ -203,3 +205,56 @@ func (i instrumentationVersionOption) ApplyTracer(config *TracerConfig) {
...
+type ErrorConfig struct {
+	Code    codes.Code
+	Message string
+	// EventOpts will be passed to addEvent
+	EventOpts []EventOption
+}
+
+// ErrorOption applies an option to a ErrorConfig.
+type ErrorOption interface {
+	ApplyError(config *ErrorConfig)
```
Comment (Aneurysm9) [inline trace/config.go:219]:
```suggestion
	applyError(config *errorConfig)
```

We're working on [normalizing configuration and options](https://github.com/open-telemetry/opentelemetry-go/pull/1882/files) and part of that will involve making the config type and apply methods unexported.  Can we get out ahead of that here?

Comment (lastchiliarch) [reply trace/config.go]: Sure. I had renamed  ApplyError to applyError.
But golangci-lint will  fail if I changed  ErrorConfig to errorConfig. 

`
exported func NewErrorConfig returns unexported type *go.opentelemetry.io/otel/trace.errorConfig, which can be annoying to us.
`

Comment (Aneurysm9) [reply trace/config.go]: `newErrorConfig` shouldn't need to be exported, either.

File: trace/config.go [hunk 10]
```diff
@@ -203,3 +205,56 @@ func (i instrumentationVersionOption) ApplyTracer(config *TracerConfig) {
...
+}
+
+// ErrorOption applies an option to a ErrorConfig.
+type ErrorOption interface {
+	applyError(config *ErrorConfig)
+
+	// A private method to prevent users implementing the
+	// interface and so future additions to it will not
+	// violate compatibility.
+	private()
```
Comment (Aneurysm9) [inline trace/config.go:224]:
```suggestion
```

This is no longer necessary since the `applyError()` method is unexported.

File: trace/config.go (hunk 9 above)
Comment (lastchiliarch) [reply trace/config.go]: Since newErrorConfig will be used in by bridge/opentracing/internal/mock.go ,  oteltest/span.go and sdk/trace/span.go, it will take a lot of work to make it unexported at this time.

I think it's better to move all the code which implements Span interface to the package trace, so we can easily make the config unexported.

But there is another problem,  since  sdk/trace/span.go needs some code in  sdk/trace/internal, we must change it at the same time.

I'm not sure about this solution, it will cause a lot of files to be changed.

Maybe we can keep NewErrorConfig exported and  make a new pr to make it unexported, I'd like to take it.

File: sdk/trace/span.go [hunk 11]
```diff
@@ -267,20 +267,25 @@ func (s *span) End(options ...trace.SpanOption) {
 	}
 }
 
-// RecordError will record err as a span event for this span. An additional call to
-// SetStatus is required if the Status of the Span should be set to Error, this method
-// does not change the Span status. If this span is not being recorded or err is nil
-// than this method does nothing.
-func (s *span) RecordError(err error, opts ...trace.EventOption) {
+// RecordError will record err as a span event for this span.
+// This method does not change the Span status by default
```
Comment (MrAlias) [inline sdk/trace/span.go:271]:
```suggestion
// This method does not change the Span status by default.
```

File: trace/config.go [hunk 12]
```diff
@@ -203,3 +205,56 @@ func (i instrumentationVersionOption) ApplyTracer(config *TracerConfig) {
 }
 
 func (instrumentationVersionOption) private() {}
+
+// ErrorConfig is a group of options for error.
```
Comment (MrAlias) [inline trace/config.go:209]:
```suggestion
// ErrorConfig is a group of options for error events.
```

File: trace/config.go (hunk 6 above)
Comment (MrAlias) [reply trace/config.go]: Why not just embed the `ErrorOption` in the `EventOption`? https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#dealing-with-overlap

**Comment by MrAlias** [comment]
Closing as HEAD has progressed far enough this will likely need to be re-evaluated.
